$OpenBSD: patch-lib_sharedbook_c,v 1.1.2.1 2010/01/02 23:12:35 william Exp $

SECURITY FIX

fix denial of service via underpopulated Huffman trees


--- lib/sharedbook.c.orig	Mon Jul 23 20:09:47 2007
+++ lib/sharedbook.c	Tue Dec  1 22:53:21 2009
@@ -124,7 +124,14 @@ ogg_uint32_t *_make_words(long *l,long n,long sparseco
     }else
       if(sparsecount==0)count++;
   }
-    
+  
+  /* sanity check the huffman tree; an underpopulated tree must be rejected. */
+  for(i=1;i<33;i++)
+    if(marker[i] & (0xffffffffUL>>(32-i))){
+      _ogg_free(r);
+      return(NULL);
+    }
+
   /* bitreverse the words because our bitwise packer/unpacker is LSb
      endian */
   for(i=0,count=0;i<n;i++){
